devito/data/data.py

```diff
@@ -365,3 +560,151 @@ def index_handle_oob(idx):
                              "multidimensional index arrays")
     else:
         return idx
+
+
+def local_data_idx(loc_idx):
+    """Retrieve the unflipped idx corresponding to loc_idx."""
+    loc_data_idx = []
+    for i in as_tuple(loc_idx):
+        if isinstance(i, slice) and i.step is not None and i.step == -1:
+            if i.stop is None:
+                loc_data_idx.append(slice(0, i.start+1, -i.step))
+            else:
+                loc_data_idx.append(slice(i.stop+1, i.start+1, -i.step))
+        elif isinstance(i, slice) and i.step is not None and i.step < -1:
+            if i.stop is None:
+                lmin = i.start
+                while lmin >= 0:
+                    lmin += i.step
+                loc_data_idx.append(slice(lmin-i.step, i.start+1, -i.step))
+            else:
+                loc_data_idx.append(slice(i.stop+1, i.start+1, -i.step))
+        elif is_integer(i):
+            loc_data_idx.append(slice(i, i+1, 1))
+        else:
+            loc_data_idx.append(i)
+    return as_tuple(loc_data_idx)
+
+
+class SLICING(Tag):
+    pass
+MPI = SLICING('mpi')  # noqa
+STANDARD = SLICING('standard')
+
+
+def mpi_index_maps(loc_idx, val_shape, topology, rank_coords, comm):
```

Comment:
this free function, and perhaps many of the other free functions from this file, should probably be moved to a different file

``rank_coords`` for homogeneity should probably just be ``coords``

how is ``val_shape`` different than ``shape``?

Reply:
`utils.py`?

Reply:
It's the `local_val.shape` which can be different from `self.shape`. The naming is a relic from when this was a class method (since 'shape' could easily be confused with `self.shape`) but since it's now a free function the `val_` can be removed.